### PR TITLE
LPS-154718 | Add tests, macros, paths for remote-apps

### PR DESCRIPTION
**Pull request by bronsonsoda**
**Test plan**
_AssertErrorMessegeAppearsWhenIFrameFieldsEmpty_
Given Add IFrame Remote Apps admin page
When required fields Name and URL are blank
And When attempt to save remote app
Then error message "field is required" appears under URL
And Then Remote App is not added

_AssertErrorMessegeAppearsWhenInvaildURL_
Given Add IFrame Remote Apps
When type testurl in URL field
And When a valid name is entered
And When attempt to save the remote app
Then error message "Please enter a valid URL" appears at the top of the page

https://issues.liferay.com/browse/LPS-163527

### CODE REVIEW

**Comment by bronsonsoda** [comment]
ci:test:frontend:nocompile

File: modules/apps/frontend-js/frontend-js-remote/tests/RemoteAppliactions.testcase [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+	tearDown {
+		var testPortalInstance = PropsUtil.get("test.portal.instance");
+
+		PortalInstances.tearDownCP();
+	}
+
+	@description = "LPS-163527. Assert iFrame is not add and warnings appear."
+	@priority = "4"
+	test AssertErrorMessegeAppearsWhenIFrameFieldsEmpty {
```
Comment (john-co) [inline modules/apps/frontend-js/frontend-js-remote/tests/RemoteAppliactions.testcase:24]: HasMandatoryFieldErrorMessagesWhenEmpty

see for more tips: https://liferay.slack.com/archives/CLD1FFVL6/p1618527187013400?thread_ts=1618516694.010400&cid=CLD1FFVL6

File: modules/apps/frontend-js/frontend-js-remote/tests/RemoteAppliactions.testcase [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+		task ("Given Add IFrame Remote Apps admin page") {
+			ApplicationsMenu.gotoPortlet(
+				category = "Custom Apps",
+				panel = "Applications",
+				portlet = "Remote Apps");
+
+			RemoteApplications.AddiFrame();
+		}
+
+		task ("When required fields Name and URL are blank and When attempt to save remote app") {
```
Comment (john-co) [inline modules/apps/frontend-js/frontend-js-remote/tests/RemoteAppliactions.testcase:34]: `When required fields Name and URL are blank` can be it's own macro to clear the fields to make sure they're blank.

File: modules/apps/frontend-js/frontend-js-remote/tests/RemoteAppliactions.testcase [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+			Button.clickPublish();
+		}
+
+		task ("And when error message "field is required" appears under URL") {
```
Comment (john-co) [inline modules/apps/frontend-js/frontend-js-remote/tests/RemoteAppliactions.testcase:38]: This should be a Then step

File: modules/apps/frontend-js/frontend-js-remote/tests/RemoteAppliactions.testcase [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+				panel = "Applications",
+				portlet = "Remote Apps");
+
+			RemoteApplications.AddiFrame();
+		}
+
+		task ("When required fields Name and URL are blank and When attempt to save remote app") {
+			Button.clickPublish();
+		}
+
```
Comment (john-co) [inline modules/apps/frontend-js/frontend-js-remote/tests/RemoteAppliactions.testcase:37]: There should be a step to assert Name field is required too right?

File: modules/apps/frontend-js/frontend-js-remote/tests/RemoteAppliactions.testcase [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+			AssertTextEquals(
+				locator1 = "Message#EMPTY_STATE_INFO",
+				value1 = "No Results Found");
+		}
+	}
+
+	@description = "LPS-163527. Assert error messege appears when invaild url is entered."
+	@priority = "4"
+	test AssertErrorMessegeAppearsWhenInvaildURL {
```
Comment (john-co) [inline modules/apps/frontend-js/frontend-js-remote/tests/RemoteAppliactions.testcase:55]: Rename to HasAnUnderstandableErrorMessageWhenURLInvalid

File: modules/apps/frontend-js/frontend-js-remote/tests/RemoteAppliactions.testcase [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+			RemoteApplications.iFrameName();
+
+			RemoteApplications.iFrameURL();
+		}
+
+		task ("And When attempt to save the remote app") {
+			Button.clickPublish();
+		}
+
+		task ("Then error message "Please enter a valid URL" appears at the top of the paged") {
```
Comment (john-co) [inline modules/apps/frontend-js/frontend-js-remote/tests/RemoteAppliactions.testcase:75]: The double quotes inside task may get into some syntax errors. Maybe use single quotes inside `'Please enter a valid URL'`

**Review by john-co — CHANGES_REQUESTED** [review]

File: modules/apps/frontend-js/frontend-js-remote/macros/RemoteApplications.macro [hunk 7]
```diff
@@ -0,0 +1,21 @@
+definition {
+
+	macro AddiFrame {
+		Button.clickAdd();
+
+		Click.clickAt(locator1 = "RemoteAppliactions#ADD_IFRAME");
+	}
```
Comment (john-co) [inline modules/apps/frontend-js/frontend-js-remote/macros/RemoteApplications.macro:7]: we have RemoteApps.addType

File: modules/apps/frontend-js/frontend-js-remote/macros/RemoteApplications.macro [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+	macro iFrameName {
+		Type(
+			key_text = "Test Name",
+			locator1 = "RemoteApplications#IFRAME_NAME");
+	}
+
+	macro iFrameURL {
+		Type(
+			key_text = "testurl",
+			locator1 = "RemoteApplications#IFRAME_URL");
```
Comment (john-co) [inline modules/apps/frontend-js/frontend-js-remote/macros/RemoteApplications.macro:18]: These may already have they're own macros as well in RemoteApps.macro

File: modules/apps/frontend-js/frontend-js-remote/paths/RemoteApplications.path [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+	<td></td>
+</tr>
+<tr>
+	<td>IFRAME_URL</td>
+	<td>//input[contains(@class,'field form-control error-field')]</td>
+	<td></td>
+</tr>
+<tr>
+	<td>PLEASE_ENTER_VALID_URL</td>
+	<td>//div[contains(@class,'alert alert-dismissible alert-danger')]</td>
```
Comment (john-co) [inline modules/apps/frontend-js/frontend-js-remote/paths/RemoteApplications.path:35]: We want the name of the path to make sense with the element xpath. Here there's no mention of Please enter valid url. The xpath just has alert, alert-dismissible and alert-danger.

If we're choosing a generic danger type alert message we can reuse Message#ERROR or something similar in Message.path file.

File: modules/apps/frontend-js/frontend-js-remote/paths/RemoteApplications.path [hunk 10]
```diff
@@ -0,0 +1,41 @@
+<html>
+<head>
+<title>RemoteApplications</title>
```
Comment (john-co) [inline modules/apps/frontend-js/frontend-js-remote/paths/RemoteApplications.path:3]: Please check RemoteApps.path for any existing xpaths to reuse

**Pull request closed without merging**